### ISSUE TITLE
Debugger does not work if `forcedebugger` is set before `super()`

### ISSUE THREAD

**Issue by IQAndreas** [FlixelCommunity/flixel#158]
Here is the thread that brings up the issue http://forums.flixel.org/index.php?topic=4385.0 (found because I ran into the same problem). I have confirmed the bug exists.

I'm sure it's a quick and simple fix, so I'll take a look at it tomorrow.

**Comment by WingEraser** [comment]
According to his last post, he put it above the super()....no wonder why it doesn't work. In the constructor the forceDebugger is set to false. It's an user error.

**Comment by Dovyski** [comment]
@WingEraser has a point, but I don't like how `forceDebugger` works. If you don't look at `FlxGame`, you will never figure out that it sets `forceDebugger` to false.

The first time I saw `forceDebugger`, I thought it would force the debug mode, no matter where I am in the code. I am pretty sure the issue author thought the same. An easy fix is to remove the `forceDebugger` from `FlxGame` constructor. It will allow the developer to place `forceDebugger` anywhere within the constructor of his game.

**Comment by WingEraser** [comment]
Hmm yeh, you want to make it static. Do you also want to apply to `useSoundHotKeys` and `useSystemCursor`?

**Comment by Dovyski** [comment]
It doesn't have to be static. `FlxGame` will instantiate the `FlxDebugger` only when the game is added to the stage, which means _after_ the constructor call. As a consequence `forceDebugger` can be placed anywhere within the constructor :)

We can extend the same idea to `useSoundHotKeys` (it works the same way). About `useSystemCursor`, it is set as a parameter in `FlxGame`, so I think it should remain the way it is.

**Comment by IQAndreas** [comment]
> The first time I saw `forceDebugger`, I thought it would force the debug mode, no matter where I am in the code. I am pretty sure the issue author thought the same.

I had the exact same problem at first! How about renaming it to something like `enableDebugger` or `allowDebugger`?

On a related topic, is there any way to open the debugger via code (so the game developer can open it, and it's not locked in any of FlxG's private or protected code), or should we add `showDebugger()` and `hideDebugger()` methods?

**Comment by Dovyski** [comment]
I guess renaming `forceDebugger` will cause a massive code break. Lots of developers will knock on our doors :)

> On a related topic, is there any way to open the debugger via code (so the game developer can open it, and it's not locked in any of FlxG's private or protected code), or should we add showDebugger() and hideDebugger() methods?

That was one of the first things I tried to find when I started working with Flixel. As far as I know, there is no way to open the debugger programmatically. It would be great to have them!  We should create a new feature request with those methods.

**Comment by IQAndreas** [comment]
> I guess renaming forceDebugger will cause a massive code break. Lots of developers will knock on our doors :)

Not necissarily. We could do something like this (using `useDebugger`, `enableDebugger`, `allowDebugger`, or whatever suits your fancy:

``` as3
public function get forceDebugger():Boolean
{ return this.enableDebugger; }
public function set forceDebugger(value:Boolean):void
{
    FlxG.log("Warning: enable the new property `enableDebugger` instead of the old `forceDebugger`!");
    this.enableDebugger = value;
}

// Initialize as false here, so it still works regardless of whether it was set before or after "super()"
private var _enableDebugger:Boolean = false;
public function get enableDebugger():Boolean
{ return _enableDebugger; }
public function set enableDebugger(value:Boolean):void
{
    _enableDebugger = value;
    if (!_enableDebugger || _currentlyShowingDebugger)
        this.hideDebugger();
}

public function showDebugger(force:Boolean = false):Boolean
{
    if (force || _enableDebugger)
    {
        // Show the debugger here...
    }
}
```

And yes, the code is quite flawed, but you get this gist. It adds the new `useDebugger` name, but still allows the old `forceDebugger` without any errors thrown (even alerting legacy developers to the change).

The `force` argument allows for these types of cases:

``` as3
function onTildePressed():void
{
    this.showDebugger();
    // Since "force" is false, it won't show the debugger if "useDebugger" is set to false.
}

function onCriticalError(error):void
{
    if (runningLocally)
    {
        showDebugger(true);
    }
    else if (betaTesting)
    {
        showDebugger(true);
        FlxG.log("Please copy and paste the above error message and send it to <admin@myflixelgame.com>!");
    }
    else
    {
        sendErrorToServer(error);
    }
}
```

**Comment by greysondn** [comment]
To chime in:

It's code design basics to make sure that /the name of the member matches what it does/. I never interpreted it to enable the debugger, but this is an interesting quandry that comes down to discovering bad design compared to expected semantic meaning - in English, "Despite seeming to be named for what it presumably does, forceDebugger clearly does not work correctly."

I don't think the answer is to rename so much as it is to find some fix that solves the perceived/semantic meaning divide. It does indeed force the debugger to exist (from a user perspective).

Bottom line: name seems right; current code is clearly wrong if it doesn't work as advertised. (Edit: as of 2.55 release or accepted community patches - not as of the suggestions and code on this issue. I've not fully reviewed all of them.)

Edits: Marked, cleanup of meaning, removal of unneeded text and confusing supplementary points.

**Comment by moly** [comment]
> An easy fix is to remove the `forceDebugger` from `FlxGame` constructor. It will allow the developer to place `forceDebugger` anywhere within the constructor of his game.

My only problem with this is that it violates flixel's existing style. In every other class I've looked at, all member variables are assigned a default value in the constructor. Not doing so here would potentially leave `forceDebugger` uninitialised.

Personally, I think `forceDebugger` should be made private and set through the `FlxGame` constructor, just like `useSystemCursor` is. That way you remove any ambiguity over where it should be used. It would break some existing code, but it's a trivial single line fix so I can't see there being many complaints. As long as it's commented correctly it should be fine.

**Comment by Dovyski** [comment]
I liked the [suggestion](https://github.com/FlixelCommunity/flixel/issues/158#issuecomment-15124379)  @IQAndreas  wrote. A name as `useDebugger` (or `enableDebugger`) makes much more sense to me (plus the `showDebugger()` is really nice!).

> moly:
> Personally, I think forceDebugger should be made private and set through the FlxGame constructor, just like useSystemCursor is. That way you remove any ambiguity over where it should be used. It would break some existing code, but it's a trivial single line fix so I can't see there being many complaints. As long as it's commented correctly it should be fine.

That makes sense too. The only thing that concerns me is the amount of parameter `FlxGame` constructor might receive in the future. It already has 7 (!) and with `useDebugger/enableDebugger` it will be 8. What if we add even more?

**Comment by IQAndreas** [comment]
Should we move this up to the next version, or include it in Flixel v2.56? If coded properly, it shouldn't break any existing code. And we could move features such as the `showDebugger()` function into a future release.

**Comment by Dovyski** [comment]
I think we should move this up to the next version, that way we can focus on the currently open bugs.

**Comment by greysondn** [comment]
Isn't this tied to some of the work in #157 that I had planned to solve via #160 once rewritten? That was just a simple refactor but I did want to see where that stands in relation to this. (I have been tied up with real life woes and waiting for the API and paths to stabilize after reorganization.)

**Comment by Dovyski** [comment]
It's somehow related, but I don't know if it is strongly connected with the refactoring. I've [been working](https://github.com/Dovyski/flixel/commit/7e9f71708b23e841cedc059548f18d43887a6312) on the new debugger and I've made some changes there regarding `forceDebugger`.

Time is a bit short for me too, but as soon as I can I'll pause my work on the debugger and merge #160. After that, I can pull and merge your changes into my debugger.

**Comment by Dovyski** [comment]
With the addition of the new debugger, `forceDebugger` is no longer necessary. It was deprecated in favor of `FlxG.debug`, which is a lot easier to understand and use.
